### PR TITLE
Referrer Policy

### PR DESCRIPTION
**Pull request by kristijanburnik**
This PR introduces the Referrer Policy testsuite covering the draft specification https://w3c.github.io/webappsec/specs/referrer-policy/
- Add referrer-policy directory for all tests regarding the Referrer Policy specification.
- Add stub for the README file on how to write these tests.
- Introduce a fundamental test: Check that a loaded subresource (e.g. an iframe) receives the referrer URL from the parent when no explicit Referrer Policy is set.

The newly added test runs successfully on x64 Linux:
- Google Chrome 42.0.2311.135 (Official Build) (64-bit)
- Mozilla Firefox 37.0.2

Other browsers and OS still untested.

### CODE REVIEW

**Comment by kristijanburnik** [comment]
@mikewest, @jeisinger 

Hi, I've added the inital referrer-policy test. 

As I'm still quite new to the web-platform-tests practices, please shout about any violations. :-)

Also, not sure how to add the test to the testsuite, since MANIFEST.json is ignored.
In any case, this should reside there:

```
{
        "path": "referrer-policy/no-referrer-policy/no-referrer-policy.html",
        "url": "/referrer-policy/no-referrer-policy/no-referrer-policy.html"
},
```

File: referrer-policy/generic/common.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
```
Comment (jeisinger) [inline referrer-policy/generic/common.js]: no empty line here

File: referrer-policy/README.html [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<!DOCTYPE html>
+<head>
+<title>Referrer-Policy Web Platform Tests - README</title>
+</head>
+<body>
+  <p>
```
Comment (jeisinger) [inline referrer-policy/README.html]: you indend html 2sp and js 4sp, is that the required style?

File: referrer-policy/generic/common.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+// Adds a PRE tag to end of body with propertis of the current page.
```
Comment (mikewest) [inline referrer-policy/generic/common.js]: Nit: s/propertis/properties/

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+       referrer URL from the child.</p>
+
+    <script>
+
+      // Display this page's location.
+      appendPageInfo("Parent frame");
+
+      // Embed the iframe on a different domain - load subresource.
+      var iframe = document.createElement('iframe');
+      iframe.type = 'text/javascript';
```
Comment (jeisinger) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: either use ' or " all the time

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE HTML>
```
Comment (jeisinger) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html]: either always HTML or always html

File: referrer-policy/generic/common.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+function appendPageInfo(title) {
```
Comment (mikewest) [inline referrer-policy/generic/common.js]: It's not entirely clear to me why you append this data to the document. Is this just for manual debugging? The actual tests should be assertion-based and automated.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 7]
```diff
@@ -0,0 +1,60 @@
+<!DOCTYPE HTML>
+<html>
+  <head>
+    <title>Check that subresource gets the referrer URL when no explicit
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: I think you'll need to be a little more structured with the titles, as there's a huge set of things we'll need to test. Something like "HTTP -> HTTP: Subresource" might make sense, and would allow "HTTP -> HTTPS: Subresource"/"HTTP -> HTTPS: Navigation", etc in the future.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 8]
```diff
@@ -0,0 +1,60 @@
+<!DOCTYPE HTML>
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Nit: s/HTML/html/

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+    <script>
+
+      // Display this page's location.
+      appendPageInfo("Parent frame");
+
+      // Embed the iframe on a different domain - load subresource.
+      var iframe = document.createElement('iframe');
+      iframe.type = 'text/javascript';
+      iframe.src = "http://www1." + location.hostname + ":" + location.port +
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Do you intend this to be cross-origin?

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 7 above)
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Now that I look below, you'll also want to include same-origin/cross-origin.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                   "/referrer-policy/no-referrer-policy" +
+                   "/no-referrer-policy.subresource.html";
+      iframe.width="100%";
+      iframe.height="200px";
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Are these relevant to the test?

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 4 above)
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: And why are you framing a `text/javascript` resource?

File: referrer-policy/README.html (hunk 2 above)
Comment (kristijanburnik) [inline referrer-policy/README.html]: All is 2sp - 4sp only for multiple line spanning statements. What should I use?

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+      document.body.appendChild(iframe);
+
+      var referrerTest = async_test(
+          "Checks that referer is passed to the subresource (iframe).");
+
+      function iFrameListener(event) {
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: You need to validate that the message is coming from the source you expect; something like `if (event.source != iframe.contentWindow) return;` should work.

File: referrer-policy/generic/common.js (hunk 1 above)
Comment (kristijanburnik) [inline referrer-policy/generic/common.js]: Done.

File: referrer-policy/generic/common.js (hunk 3 above)
Comment (kristijanburnik) [inline referrer-policy/generic/common.js]: Done.

File: referrer-policy/generic/common.js (hunk 6 above)
Comment (kristijanburnik) [inline referrer-policy/generic/common.js]: This is only for preview purpose if you run the test manually. The test works automated.
Do you suggest I should remove all meta for debugging?

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        referrerTest.step(function() {
+            // Sanity check - location of the iframe matches reported location.
+            assert_equals(event.data.location, iframe.src,
+                          "Check child reported location.");
+
+            // Child reports the referrer URL.
+            assert_equals(event.data.referrer, document.location.toString(),
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: This isn't quite right, at least not in the general case. We strip fragment identifiers from referrers, which we should eventually have a test for. It might be reasonable to add a `stripForUseAsReferrer` function to your utility JavaScript file which implements https://w3c.github.io/webappsec/specs/referrer-policy/#strip-url for this comparison.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 4 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Legacy from copying. Removed.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 9 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Yes. This is explicitly to test cross-origin referrer. Should I point it out in a comment?

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html (hunk 5 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html]: Done.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 7 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Ok. Adding another iframe for the samge-origin URL. Will update PR shortly.

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  <body style="background:#f4f4f4;">
+    <h2>This page reports back it's referrer URL to the parent frame.</h2>
+
+    <script>
+      // Display this page's location (iframe page - subresource to parent).
+      appendPageInfo("Child frame");
+
+      var pageInfo = {
+          location:document.location.toString(),
+          referrer:document.referrer
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html]: Note that this test only verifies the JavaScript interface. You could convert it to a Python file which outputs the HTTP header as well (e.g. `print "<script>var httpReferer = '%s';</script>" % request.headers["referer"]` (see something like https://github.com/w3c/web-platform-tests/blob/master/eventsource/resources/cors-cookie.py for something vaguely similar).

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 9 above)
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Yup! :)

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 7 above)
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: I'd suggest not adding another frame, but splitting it out into a separate test. Multiple frames introduces flakiness (Maybe frame 1 loads first? Maybe frame 2 hangs for an hour?), single-frame tests are clearer and test only a single thing. That's a good goal to aim for.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 11 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: I have refactored this so that each iframe receives and ID as a query parameter. The iframe then sends back that ID along with the referrer. I then reuse that ID to run the appropriate test scenario. Also added some sanity checks.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 12 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: I see.. I have added this helper function to generic/common.js. 
Currently only to strip the fragment.  Added a TODO since other tests will most certainly run with fragments, username, password etc. 

Question:
If I implement this function, shouldn't I write a test for it as well? :-)

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    <h2>This page reports back it's referrer URL to the parent frame.</h2>
+
+    <script>
+      // Display this page's location (iframe page - subresource to parent).
+      appendPageInfo("Child frame");
+
+      // Note: Read the ID from the query section of this page's URL.
+      // The ID is used to identify the sender of the message in the parent.
+      var pageInfo = {
+          id: document.location.search.replace("?id=", ""),
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html]: This is a bit fragile, as it will break if `id` is not the only parameter. It's probably fine for now, but perhaps adding a `TODO` would be appropriate.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+    <p>This page loads sub-resources (iframes) and waits for messages containing
+       referrer URLs from children.</p>
+
+    <script>
+      // The referrer URL should be this document's URL adjusted according to
+      // this section:
+      // https://w3c.github.io/webappsec/specs/referrer-policy/#strip-url
+      var expectedReferrer =
+          stripUrlForUseAsReferrer(document.location.toString());
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Nit: 80 columns is lame. :)

**Comment by mikewest** [comment]
I'm happy to land this for you if you're planning on checking the actual HTTP header in a followup patch. Is that your plan?

**Comment by kristijanburnik** [comment]
I'm still working on nits.

I might even add the HTTP header checks as you suggested for review before
landing it.

On Wed, May 6, 2015 at 4:21 PM, Mike West <notifications@github.com> wrote:

> I'm happy to land this for you if you're planning on checking the actual
> HTTP header in a followup patch. Is that your plan?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/1807#issuecomment-99485857>
> .
>



-- 

*Kristijan Burnik*

Software Engineer Intern

burnik@google.com
+385 91 4846 379

Google Germany GmbH

Dienerstraße 12

80331 München

Geschäftsführer: Graham Law, Christine Elizabeth Flores

Registergericht und -nummer: Hamburg, HRB 86891

Sitz der Gesellschaft: Hamburg

Diese E-Mail ist vertraulich. Wenn Sie nicht der richtige Adressat sind,
leiten Sie diese bitte nicht weiter, informieren Sie den Absender und
löschen Sie die E-Mail und alle Anhänge. Vielen Dank.



This e-mail is confidential. If you are not the right addressee please do
not forward it, please inform the sender, and please erase this e-mail
including any attachments. Thanks.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 15 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Can we settle for 120? :-)

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 7 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Just to follow up on the current implementation. I'm actually using multiple frames - one per scenario. The order in which they load does not matter since each of them receive an ID and the message listener is mapping the appropriate test to run via that ID.  Should running independent tests in a particular order ever matter? As for the hanging, is it more likely to have an iframe hang if it's on the same page of the same test or if it's loaded once per test html file (popup)?

I suppose it would be a good idea to split the test to have only one per file (html page). However, that could be done easily if I continue to introduce more structure into the tests and as that structure emerges. At this point it's easier to keep them in one place while I'm still thinking up new scenarios for the referrer-policy and as I read more of the specs.

I plan to do plenty of refactoring for reusability as things evolve.

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html (hunk 13 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html]: Yes, I agree, I think it would be good if I serialized all the headers via python to JS and provide them to the subresource - this can be reused in multiple tests.

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html (hunk 14 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.html]: Fixed via parsing the query string.

**Comment by kristijanburnik** [comment]
Implemented checking request headers. :-)

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+      // Map all test arrangements by ID. With this mapping loop we also append
+      // the iframes to the body.
+      for (var i in scenarios) {
+        var scenario = scenarios[i];
+        testArrangementMap[scenario.id] =
+            arrangeReferrerPolicyAsyncTest(scenario.id,
+                                           scenario.url,
+                                           scenario.description);
+      }
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: I think we'll end up wanting to move some substantial subset of this code out into generic methods in the common JS file as you add more and more tests. There's no reason to do that right now, but keep it in mind for future patches.

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py [hunk 17]
```diff
@@ -0,0 +1,16 @@
+import os, json
+
+def main(request, response):
+    script_directory = os.path.dirname(os.path.abspath(__file__))
+    template_basename = "no-referrer-policy.subresource.template.html"
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py]: Likewise, I think you'll want to turn this into a more generic loading mechanism rather than spawning a variety of .py files. Again, something to consider in future patches.

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+    template_filename = os.path.join(script_directory, template_basename);
+
+    with open(template_filename) as f:
+        template = f.read()
+
+    headers_as_json = json.dumps(request.headers)
+    replacement_tag = "/* <SERVER_REQUEST_HEADERS> */"
+    exported_headers = "var SERVER_REQUEST_HEADERS = " + headers_as_json + ";"
+    rendered_html = template.replace(replacement_tag, exported_headers)
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py]: Hrm. This is a bit ad hoc. It's fine for right now, but I think it would be better to use a real templating library, rather than poking at the file with a replacement method.

Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py]: (In a future patch)

Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py]: Actually, I take it back: I'd like you to change this to a named formatting variable. That is, please add something like `%(headers)` to the templated HTML file, and do the replacement as `template % { 'headers': headers_as_json }`.

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.template.html [hunk 19]
```diff
@@ -0,0 +1,34 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>This page reports back it's request details the parent frame</title>
+    <!-- Common global functions for referrer-policy tests. -->
+    <script src="../generic/common.js"></script>
+  </head>
+  <body style="background:#f4f4f4;">
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.template.html]: 1. Why `#f4f4f4`?
2. Why not do this via `<style>body { background ... }</style>`?

File: referrer-policy/no-referrer-policy/no-referrer-policy.html [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+      var documentUrl = document.location.toString();
+
+      // The referrer URL should be this document's URL adjusted according to
+      // this section:
+      // https://w3c.github.io/webappsec/specs/referrer-policy/#strip-url
+      var expectedReferrerUrl = stripUrlForUseAsReferrer(documentUrl);
+
+      // Path to the sub-resource page which feeds back received referrer URL.
+      var subresourceUrlPath = "/referrer-policy/no-referrer-policy" +
+                               "/no-referrer-policy.subresource.py";
```
Comment (mikewest) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: In a future patch, I suspect you'll want to make these parameters to some sort of assertion so you can reuse some of this code elsewhere.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 16 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: Agreed. I usually tend not to overgeneralize and overengineer in the start - when I have only one instance of something (e.g. only one spec case) I keep it simple - I think it's easier for review as well. Once I add a second instance I extract the commonalities for reuse as they naturally present themselves and then refactor.

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py (hunk 17 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py]: I was thinking the same thing as I was repushing. Actually I think one python script should be sufficient for the whole referrer policy. As this whole thing is a test in itself, such refactoring should come easy and without fear of breaking stuff.

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py (hunk 18 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.py]: Done: %(headers)s :-)

File: referrer-policy/no-referrer-policy/no-referrer-policy.subresource.template.html (hunk 19 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.subresource.template.html]: Removed alltogether. Looked nice though when there was one iframe.

File: referrer-policy/no-referrer-policy/no-referrer-policy.html (hunk 20 above)
Comment (kristijanburnik) [inline referrer-policy/no-referrer-policy/no-referrer-policy.html]: This is already reused in the tests below. Once I add more scenarios I'll rethink how to separate them and reuse a bunch of stuff.

**Comment by kristijanburnik** [comment]
No problem. Thanks. :-)

On Thu, May 7, 2015 at 7:58 PM, Mike West <notifications@github.com> wrote:

> Ok, this looks like a fine first pass. Would you mind squashing this into
> a single commit so I can land it for you without all the nits and updates?
> :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/1807#issuecomment-99961175>
> .
>



-- 

*Kristijan Burnik*

Software Engineer Intern

burnik@google.com
+385 91 4846 379

Google Germany GmbH

Dienerstraße 12

80331 München

Geschäftsführer: Graham Law, Christine Elizabeth Flores

Registergericht und -nummer: Hamburg, HRB 86891

Sitz der Gesellschaft: Hamburg

Diese E-Mail ist vertraulich. Wenn Sie nicht der richtige Adressat sind,
leiten Sie diese bitte nicht weiter, informieren Sie den Absender und
löschen Sie die E-Mail und alle Anhänge. Vielen Dank.



This e-mail is confidential. If you are not the right addressee please do
not forward it, please inform the sender, and please erase this e-mail
including any attachments. Thanks.

**Comment by mikewest** [comment]
LGTM, thanks for taking a few passes on this.

**Comment by mikewest** [comment]
I'll wait till Travis runs lint before landing.

**Comment by kristijanburnik** [comment]
Woohoo! Awesome. :-)

On Thu, May 7, 2015 at 9:24 PM, Mike West <notifications@github.com> wrote:

> Merged #1807 <https://github.com/w3c/web-platform-tests/pull/1807>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/1807#event-299609576>.
>



-- 

*Kristijan Burnik*

Software Engineer Intern

burnik@google.com
+385 91 4846 379

Google Germany GmbH

Dienerstraße 12

80331 München

Geschäftsführer: Graham Law, Christine Elizabeth Flores

Registergericht und -nummer: Hamburg, HRB 86891

Sitz der Gesellschaft: Hamburg

Diese E-Mail ist vertraulich. Wenn Sie nicht der richtige Adressat sind,
leiten Sie diese bitte nicht weiter, informieren Sie den Absender und
löschen Sie die E-Mail und alle Anhänge. Vielen Dank.



This e-mail is confidential. If you are not the right addressee please do
not forward it, please inform the sender, and please erase this e-mail
including any attachments. Thanks.
